packages/common/infra/src/di/core/collection.ts

```diff
@@ -0,0 +1,459 @@
+import { parseIdentifier } from './identifier';
+import type { ServiceProvider } from './provider';
+import { BasicServiceProvider } from './provider';
+import { stringifyScope } from './scope';
+import type {
+  GeneralServiceIdentifier,
+  ServiceFactory,
+  ServiceIdentifier,
+  ServiceIdentifierType,
+  ServiceIdentifierValue,
+  ServiceScope,
+  ServiceVariant,
+  Type,
+  TypesToDeps,
+} from './types';
+
+/**
+ * A collection of services.
+ *
+ * ServiceCollection basically is a tuple of `[scope, identifier, variant, factory]` with some helper methods.
+ * It just stores the definitions of services. It never holds any instances of services.
+ *
+ * # Usage
+ *
+ * ```ts
+ * const services = new ServiceCollection();
+ * class ServiceA {
+ *   // ...
+ * }
+ * // add a service
+ * services.add(ServiceA);
+ *
+ * class ServiceB {
+ *   constructor(serviceA: ServiceA) {}
+ * }
+ * // add a service with dependency
+ * services.add(ServiceB, [ServiceA]);
+ *                         ^ dependency class/identifier, match ServiceB's constructor
+ *
+ * const FeatureA = createIdentifier<FeatureA>('Config');
+ *
+ * // add a implementation for a service identifier
+ * services.addImpl(FeatureA, ServiceA);
+ *
+ * // override a service
+ * services.override(ServiceA, NewServiceA);
+ *
+ * // create a service provider
+ * const provider = services.provider();
+ * ```
+ *
+ * # The data structure
+ *
+ * The data structure of ServiceCollection is a three-layer nested Map, used to represent the tuple of
+ * `[scope, identifier, variant, factory]`.
+ * Such a data structure ensures that a service factory can be uniquely determined by `[scope, identifier, variant]`.
+ *
+ * When a service added:
+ *
+ * ```ts
+ * services.add(ServiceClass)
+ * ```
+ *
+ * The data structure will be:
+ *
+ * ```ts
+ * Map {
+ *  '': Map {                      // scope
+ *   'ServiceClass': Map {         // identifier
+ *     'default':                  // variant
+ *        () => new ServiceClass() // factory
+ *  }
+ * }
+ * ```
+ *
+ * # Dependency relationship
+ *
+ * The dependency relationships of services are not actually stored in the ServiceCollection,
+ * but are transformed into a factory function when the service is added.
+ *
+ * For example:
+ *
+ * ```ts
+ * services.add(ServiceB, [ServiceA]);
+ *
+ * // is equivalent to
+ * services.addFactory(ServiceB, (provider) => new ServiceB(provider.get(ServiceA)));
+ * ```
+ *
+ * For multiple implementations of the same service identifier, can be defined as:
+ *
+ * ```ts
+ * services.add(ServiceB, [[FeatureA]]);
+ *
+ * // is equivalent to
+ * services.addFactory(ServiceB, (provider) => new ServiceB(provider.getAll(FeatureA)));
+ * ```
+ */
+export class ServiceCollection {
+  private readonly services: Map<
+    string,
+    Map<string, Map<ServiceVariant, ServiceFactory>>
+  > = new Map();
+
+  /**
+   * Create an empty service collection.
+   *
+   * same as `new ServiceCollection()`
+   */
+  static get EMPTY() {
```

Comment:
it looks dangerous to me to return a new instance on access. Maybe cache the returned value?

Reply:
What if someone modified it?

Reply:
I guess I fell into the immutable mindset now but the design of DI is built upon mutable structures.
Resolving now.